doc/continuous_integration/dockerfile.md

```diff
@@ -0,0 +1,120 @@
+# Dockerfile Documentation
+
+Dockerfiles, denoted via the `(<name>.)Dockerfile` file name extension, provides repeatable and reproducible means to build and test the project, as well as build images for running container based CI services. Further references on writing and building Dockerfiles, such as syntax and tooling can be found here:
+
+* [Dockerfile reference](https://docs.docker.com/engine/reference/builder)
+* [Best practices for writing Dockerfiles](https://docs.docker.com/develop/develop-images/dockerfile_best-practices)
+
+The Dockerfiles for this project are built upon parent images from upstream repos on DockerHub, thus abbreviating environmental setup and build time, yet written in a parameterized style to remain ROS2 distro agnostic. This keeps them easily generalizable for future ROS2 releases or for switching between custom parent images. When choosing the parent image, a tradeoff may persist between choosing a larger tag with more than what you need pre-installed, saving time building the image locally, vs. choosing a smaller tag without anything you don't need, saving time pulling or pushing the image remotely. Given the use of multiple build stages, they're consequently best approached by reading from top to bottom in the order in which image layers are appended. More info on upstream repos on DockerHub can be found here:
+
+* [ROS Docker Images](https://hub.docker.com/_/ros)
+  * DockerHub repo for official images
+* [ROS Dockerfiles](https://github.com/osrf/docker_images)
+  * GitHub repo for OSRF Dockerfiles
+* [Official Images on Docker Hub](https://docs.docker.com/docker-hub/official_images)
+
+While the main [`Dockerfile`](/Dockerfile) at the root of the repo is used for development and continuous integration, the [`.dockerhub/`](/.dockerhub) directory contains additional Dockerfiles that can be used for building the project entirely from scratch, include the minimal spanning set of recursive ROS2 dependencies from source, or building the project from a released ROS2 distro using available pre-built binary dependencies. These are particularly helpful for developers needing to build/test the project using a custom ROS2 branch, or for a users building with an alternate ROS2 base image, but are not used for the CI pipeline. We'll walk through the main Dockerfile here, although all of them follow the same basic pattern.
+
+## Global Arguments
+
+The Dockerfile first declares a number of optional `ARG` values and respective defaults to specify the parent image to build `FROM` and workspace paths. Here the Dockerfiles assume all workspaces are nested within the `/opt` directory. These `ARG`s can be accessed similarly to `ENV`s, but must be declared in a stage's scope before they can be used, and unlike `ENV` only exist at build time of that stage and do not persist in the resulting image. For multi-stage builds, the last stage is what is tagged as the final image. More info on multi-stage builds can be found here: 
```

Comment:
> I don't see the reason for this first paragraph or the multi-stage builds reference.

It explains why the `ARG` directives must be repeated for each multistage, and how despite this, we can keep the Dockerfile DRY by only declaring default values of the `ARG`s in the beginning scope. If you read the reference, it's quite relevant here.

The other non global `ARG`s are brought up when they are relevant to subsection; mixin `ARG`s in the `Build Source` section, test `ARG`s in the `Testing Overlay` section.

e4677ef949cb3b28dda6ca74ead3ab83cb9ec087

Reply:
I think if you call the section "global argument" you need to go over briefly what all the global arguments are. Even a bullet list with a sentence explanation would be enough. Otherwise, this section isn't really talking at all about global arguments.

I still think the first paragraph doesn't really aid in the explanation much. Maybe it could be simplified but as written its very abstract and I still don't see a clear connection.